File: argon-plugin/shared/src/main/scala/dev/argon/plugin/SerializedTube.scala
package dev.argon.plugin

import zio.*
import zio.stream.*

trait SerializedTube[R, E] {

  def metadata(): ZIO[R, E, dev.argon.tube.Metadata.V1]

  def moduleDeclaration(id: BigInt): ZIO[R, E, dev.argon.tube.ModuleDeclaration.V1]
  def moduleDefinition(id: BigInt): ZIO[R, E, dev.argon.tube.ModuleDefinition.V1]


  def traitRef(id: BigInt): ZIO[R, E, dev.argon.tube.TraitReference.V1]
  def traitDef(id: BigInt): ZIO[R, E, dev.argon.tube.TraitDefinition.V1]

  def classRef(id: BigInt): ZIO[R, E, dev.argon.tube.ClassReference.V1]
  def classDef(id: BigInt): ZIO[R, E, dev.argon.tube.ClassDefinition.V1]

  def functionRef(id: BigInt): ZIO[R, E, dev.argon.tube.FunctionReference.V1]
  def functionDef(id: BigInt): ZIO[R, E, dev.argon.tube.FunctionDefinition.V1]
  def externFunctionImplementation(id: BigInt, platform: Platform[R, E]): ZIO[R, E, platform.ExternFunction]

  def methodRef(id: BigInt): ZIO[R, E, dev.argon.tube.MethodReference.V1]
  def methodDef(id: BigInt): ZIO[R, E, dev.argon.tube.MethodDefinition.V1]
  def externMethodImplementation(id: BigInt, platform: Platform[R, E]): ZIO[R, E, platform.ExternMethod]

  def classConstructorRef(id: BigInt): ZIO[R, E, dev.argon.tube.ClassConstructorReference.V1]
  def classConstructorDef(id: BigInt): ZIO[R, E, dev.argon.tube.ClassConstructorDefinition.V1]
  def externClassConstructorImplementation(id: BigInt, platform: Platform[R, E]): ZIO[R, E, platform.ExternClassConstructor]
}

File: plugins/java/api/src/dev/argon/plugin/api/tube/SerializedTube.java
package dev.argon.plugin.api.tube;

import java.math.BigInteger;
import dev.argon.plugin.api.*;
import java.io.IOException;

public interface SerializedTube<E extends Exception> {
    Metadata.V1 metadata() throws E, IOException, PluginException;

    ModuleDeclaration.V1 moduleDeclaration(BigInteger id) throws E, IOException, PluginException;
    ModuleDefinition.V1 moduleDefinition(BigInteger id) throws E, IOException, PluginException;


    TraitReference.V1 traitRef(BigInteger id) throws E, IOException, PluginException;
    TraitDefinition.V1 traitDef(BigInteger id) throws E, IOException, PluginException;

    ClassReference.V1 classRef(BigInteger id) throws E, IOException, PluginException;
    ClassDefinition.V1 classDef(BigInteger id) throws E, IOException, PluginException;

    FunctionReference.V1 functionRef(BigInteger id) throws E, IOException, PluginException;
    FunctionDefinition.V1 functionDef(BigInteger id) throws E, IOException, PluginException;
    <ExternFunction> ExternFunction externFunctionImplementation(BigInteger id, Platform<E, ExternFunction, ?, ?> platform) throws E, IOException, PluginException;

    MethodReference.V1 methodRef(BigInteger id) throws E, IOException, PluginException;
    MethodDefinition.V1 methodDef(BigInteger id) throws E, IOException, PluginException;
    <ExternMethod> ExternMethod externMethodImplementation(BigInteger id, Platform<E, ?, ExternMethod, ?> platform) throws E, IOException, PluginException;

    ClassConstructorReference.V1 classConstructorRef(BigInteger id) throws E, IOException, PluginException;
    ClassConstructorDefinition.V1 classConstructorDef(BigInteger id) throws E, IOException, PluginException;
    <ExternClassCtor> ExternClassCtor externClassConstructorImplementation(BigInteger id, Platform<E, ?, ?, ExternClassCtor> platform) throws E, IOException, PluginException;
    
}
